### ISSUE TITLE
Mobile SDK blocks the UI thread when loading MBTiles in UWP Xamarin Forms apps

### ISSUE THREAD

**Issue by Shtong** [CartoDB/mobile-sdk#114]
Hello,

I've stumbled on a weird bug using the mobile SDK MapView that seems to be happening only if these conditions are met:

- The map is hosted in a Xamarin Forms UI
- The map runs in a UWP app
- The map contains a layer displaying the contents of an MBTiles database

Under these circumstances, at the moment the map is added to the UI, the UI thread hangs for 5~10 seconds. Note however that this does not happen every time (I'd say 4/5 times in my test project).

A workaround I found is to make sure the layers are created *after* the map has finished its initial render. For example:

```c#
Map = new MapView();

var startTimer = new System.Threading.Timer((_) =>
{
    Device.BeginInvokeOnMainThread(() =>
    {
      // Configure / add the MBTiles layer
    });
}, null, 10, System.Threading.Timeout.Infinite);

Content = Map.ToView();

```

**Comment by mtehver** [comment]
One idea what is happening: SDK tries to detect min/maxzoom of the tiles of MBTiles file in MBTilesTileDataSource constructor. If the file is large, it may take a while. If that is the case, then there are 2 workarounds: 1) explictly specify min/max zoom of MBTiles file in MBTilesTileDataSource constructor 2) create the datasource/layer in the background thread and add it to the mapview (SDK is mostly threadsafe, so can be done in the background thread).

**Comment by Shtong** [comment]
I don't think the performance problem is located in the MBTilesTileDataSource constructor (unless the file reading process is asynchronous and the constructor only starts it), because I timed the layer creation performance (creating the datasource, the style set, the decoder, the layer, and adding the layer to the map) and it never takes more than 300ms with the Roma data files that's in the Android sample projects. Also:

- The UI thread blocks **after** the layers have been created and added to the map
- When running the same test on an Android phone with Xamarin Forms, or on UWP without Xamarin Forms, the problem does not appear (if it was really an IO performance problem I'd expect a better performance on my PC than on my phone, and the same on my PC with or without Forms).

Because of the seemingly random occurrence of this bug, (it occurs most of the time, but not every time), I was thinking maybe of a threading related problem between Carto and the Xamarin forms UI system ?

**Comment by mtehver** [comment]
Yes, this is strange. Perhaps you could make a working sample app demonstrating the issue, so we could investigate this further? Leaving the issue open currently.

**Comment by Shtong** [comment]
I've put the project I used to investigate this problem on GitHub. It can be found here: https://github.com/Shtong/ReproCartoPerfForms

Most of the interesting logic can be found in MapContainer.cs, which is pretty straightforward.

**Comment by Nikituh** [comment]
Hey

I looked into it a bit. There does indeed seem to be a problem, but, oddly enough, it's a conflict between zoom and layer rendering. If you animate zoom, or don't zoom, there are no issues. We'll definitely look into it.

I changed your async logic a bit, as I couldn't get the project to run as is, implemented event handlers. Here's my code:

```
    public class MapContainer : ContentView
    {
        public MapView Map { get; private set; }

        public App Application { get { return ((App)App.Current); } }

        public MapContainer()
        {
            Map = new MapView();
            Content = Map.ToView();
        
            Application.FileCopied += OnFileCopied;
        }

        private void OnFileCopied(object sender, EventArgs e)
        {
            string file = (string)sender;
            MBTilesTileDataSource testDataSource = new MBTilesTileDataSource(0, 14, file);

            CompiledStyleSet styleSet = new CompiledStyleSet(new ZippedAssetPackage(AssetUtils.LoadAsset("nutibright-v3.zip")));
            MBVectorTileDecoder decoder = new MBVectorTileDecoder(styleSet);
            VectorTileLayer testLayer = new VectorTileLayer(testDataSource, decoder);

            Device.BeginInvokeOnMainThread(() =>
                {
                    MapPos pos = Map.Options.BaseProjection.FromWgs84(new MapPos(App.CITY_POS_X, App.CITY_POS_Y));
                    Map.SetFocusPos(pos, 0);
                    Map.SetZoom(15, 0.1f);
                    
                    Map.Layers.Add(testLayer);
                    Debug.WriteLine("Done!");
                });
        }
    }
```

**Comment by mtehver** [comment]
The issue is "fixed" in SDK 4.1 snapshots. There are no changes in 4.1 that should have affected this, but most likely explanation is that the issue was really fixed by moving the SDK component build system from VS2015 to VS2017 and it was related somehow to the VC runtime upgrade.

**Comment by mtehver** [comment]
This should be fixed with the latest 4.1RC3 release. Closing.
